aiida/tools/visualization/graph.py

```diff
@@ -682,11 +708,17 @@ def recurse_descendants(
 
         # Pop the origin node and add it to the graph, applying custom styling
         origin_node = traversed_nodes.pop(origin_pk)
+        if not origin_style:
+            origin_style = self._origin_node_style
```

Comment:
Hmmm.... I did this because pylint complain about using dict as default value. Maybe user should pass `{}` as the orgin style to achieve the purpose you mentioned?

Reply:
Ah, ok, I see, this is because it is not a good idea to use mutables as defaults. I think if you wrap your dictionary around `dict( your_dict )` that should do the trick.

Reply:
You could use ``types.MappingProxyType({})`` as a default value.

The reason pylint is complaining here is that having mutable defaults is dangerous - any mutation to the default object will affect subsequent function calls. The ``MappingProxyType`` wraps a dict in an immutable proxy. Since no one holds a "normal" reference to that dict (it's constructed right there), it's effectively immutable.

Reply:
> I think if you wrap your dictionary around `dict( your_dict )` that should do the trick.

@greschd Isn't this enought?

Reply:
In this specific case it would have the same effect, yes. The difference would be that wrapping it in ``dict`` just makes a copy _inside_ the function body. If you just look at the signature, you can't be sure the original dict isn't modified.

In general, you should also note that ``dict(your_dict)`` only makes a _shallow_ copy. If you can modify nested contents it will still propagate, e.g.
```python
In [1]: x = {'a': {'b': 1}}

In [2]: y = dict(x)

In [3]: y['b'] = 1

In [4]: x
Out[4]: {'a': {'b': 1}}

In [5]: y['a']['b'] = 2

In [6]: x
Out[6]: {'a': {'b': 2}}
```

The ``MappingProxyType`` has the same problem, but ``MappingProxyType({})`` specifically does not - so it is a valid for providing an empty dictionary as default.

Reply:
> I think if you wrap your dictionary around dict( your_dict ) that should do the trick.

@ramirezfranciscof That not did the trick after test still the same pylint error. 

However, `MappingProxyType` also not working here. Since `origin_style` will be passed to `add_node` method which need to update this dict:
```
        node = self._load_node(node)                                                                                    
        style = {} if style_override is None else style_override                                                        
>       style.update(self._global_node_style)                                                                           
E       AttributeError: 'mappingproxy' object has no attribute 'update'       
```

Reply:
Aha, the `MappingProxyType` actually did its job here: it's immutable, so you can not `update` it. Note that if you were to update it, it would _also_ change the default value.

So we need to combine the two solutions: Use `MappingProxyType` and 
```python
style = {} if style_override is None else dict(style_override)
```
to cast it to a mutable dict.

Reply:
I see. thanks! @greschd

Reply:
One point that I want to stress: This `MappingProxyType` approach works **only** if the values of the contained dict are all immutable. 

That is the case
```python
In [1]: from types import MappingProxyType

In [2]: x = MappingProxyType({'a': 1})

In [3]: x['a'] = 2
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-3-ed6b9b75d0cd> in <module>()
----> 1 x['a'] = 2

TypeError: 'mappingproxy' object does not support item assignment
```

If the values are themselves mutable, you can mutate them still:
```python
In [4]: x = MappingProxyType({'a': {'b': 1}})

In [5]: x['a']['b'] = 2

In [6]: x
Out[6]: mappingproxy({'a': {'b': 2}})
```

That is because only the outermost object is immutable - same as e.g. having a ``list`` inside a ``tuple`` is also mutable.

For our case here, someone could change the defaults by mutating _directly_ the ``OVERRIDE_STYLES_DICT``, like 
```python
In [1]: OVERRIDE_STYLES_DICT = {
   ...:     'ignore_node': {
   ...:         'color': 'lightgray',
   ...:         'fillcolor': 'white',
   ...:         'penwidth': 2,
   ...:     },
   ...:     'origin_node': {
   ...:         'color': 'red',
   ...:         'penwidth': 6,
   ...:     },
   ...: }

In [2]: def func(style=MappingProxyType(OVERRIDE_STYLES_DICT['origin_node'])):
   ...:     print(style)
   ...:

In [3]: func()
{'color': 'red', 'penwidth': 6}

In [4]: OVERRIDE_STYLES_DICT['origin_node']['color'] = 'blue'

In [5]: func()
{'color': 'blue', 'penwidth': 6}
```

I think this ok in that particular case though, because the capitalization of ``OVERRIDE_STYLES_DICT`` also indicates that it's supposed to be constant (and thus shouldn't be modified).

Reply:
It is strange that pylint complain about 
```
pylint: no-name-in-module / No name 'MappingProxyType' in module 'types'
```
But it fine in my local machine. All versions of pre-commit tools is same as CI test.

Reply:
I have no idea why is this; @greschd any idea? I'm running the pre-commits locally to see if something comes up.

Reply:
Uh, no idea.. `MappingProxyType` was added to `types` in Python 3.3. 

@unkcpz is the `pylint` version you've got installed locally exactly the same as what's in the `aiida-core` `setup.json`?

I'd be tempted to just slap a ``# pylint: disable=no-name-in-module,useless-suppression`` on that.

Reply:
Oh, it could be that ``aiida/params/types`` is shadowing the ``types`` module, or ``pylint`` doesn't understand that that is not a top-level module.

Reply:
Yes, the `pylint` version is same, also version of packages in "dev_precommit". @greschd

Reply:
@unkcpz We can't seem to figure it out. For now, I would say just add the `pylint: disable` and open up an issue with this problem once the PR is merged.

Reply:
Thanks. @ramirezfranciscof